### DISCUSSION TITLE
Withdraw function problem

### DISCUSSION THREAD

**Discussion by meitedaf** [Cyfrin/foundry-full-course-cu#850] [Q&A] [answered]
### Issue Description:
```
function withdraw() public onlyOwner {
        for (uint256 funderIndex=0; funderIndex < s_funders.length; funderIndex++){
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
```
In the withdraw function of the contract, certain actions are taken:

The s_funders array is cleared, which contains the addresses of all users who have previously called the fund function.

The contribution amounts of all user addresses in the s_addressToAmountFunded mapping are set to 0.

This means that when a user calls the withdraw function, the s_funders array no longer contains any addresses, and the contribution amounts of all users are reset to zero. However, the issue is that this operation only withdraws the funds of the current user who triggered the withdraw function.

### Issue Explanation:
The withdraw function seems to have an unintended side effect. While it correctly resets the contribution amounts and clears the list of funders, it only processes the withdrawal for the user who called the function. Other users' contributions and addresses are cleared from these data structures, even though their funds are not being withdrawn.

### Expected Behavior:
The expected behavior would be to have a withdraw function that allows each user to individually withdraw their contributed funds while keeping the records of other users intact. The current behavior effectively wipes out records of all users except the one initiating the withdrawal.

### Proposed Solution:
A possible solution would be to refactor the withdraw function to enable users to withdraw their respective contributions without affecting other users' data. This could involve iterating through the s_funders array and the s_addressToAmountFunded mapping to identify and process the withdrawal for the user who called the function, leaving other users' data intact.

**Comment by theirrationalone** [comment]
Heyy @meitedaf !, Actually that's an intended behavior, Designed by Patrick himself. Think it as a donation Seeking Protocol/Contract. I needed money for some financial reasons and i'm even not able to collect expected money so i made an appeal to donate money on WEB3. ;`)

**Comment by 0xusmanf** [comment, marked as answer]
> ### Issue Explanation:
> The withdraw function seems to have an unintended side effect. While it correctly resets the contribution amounts and clears the list of funders, it only processes the withdrawal for the user who called the function. Other users' contributions and addresses are cleared from these data structures, even though their funds are not being withdrawn.

The following line in the `withdraw()` function not just withdraws for the user who called the function but the total balance funded to the contract
```
(bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
```
`address(this)` means the address of the contract where the function is being executed and `address(this).balance` means the total ether balance of the contract.

> ### Expected Behavior:
> The expected behavior would be to have a withdraw function that allows each user to individually withdraw their contributed funds while keeping the records of other users intact. The current behavior effectively wipes out records of all users except the one initiating the withdrawal.

Basically, users are not funding the contract to withdraw their funds later on, funds are for the owner of the contract which is why the `withdraw` function has `onlyOwner` modifier, which means only the owner of the contract can withdraw the funds funded to the contract. 

Other users shouldn't be able to call this function as it will defeat the purpose of funding the owner of the contract and anybody would call the withdraw function to steal the funds which are meant for the owner.

As far as the resetting of the funders list and their contributions are concerned, I think it totally depends on what Patrick had in mind while creating this project. Maybe he thought of keeping this data for only active funding session and resetting it for a new session after the owner withdraws the funds.

**Reply by theirrationalone** [reply]
@usmanfarooq91 !, You nailed it 💥

**Reply by meitedaf** [reply]
Thank you! Now I get it. Your support and expertise were invaluable to me.
